src/sidebar/services/groups.js

```diff
@@ -225,75 +234,68 @@ function groups(
-                ? featuredGroups.concat([selectedGroup])
-                : featuredGroups;
-
-            // If there's a selected annotation it may require an extra api call
-            // to fetch its group.
-            if (selectedAnn) {
-              // Set the directLinkedAnnotationGroupId to be used later in
-              // the filterGroups method.
-              directLinkedAnnotationGroupId = selectedAnn.group;
-
-              const selectedAnnGroup = myGroups
-                .concat(allFeaturedGroups)
-                .some(g => g.id === selectedAnn.group);
-
-              // If the direct-linked annotation's group has not already been fetched,
-              // fetch it.
-              if (!selectedAnnGroup) {
-                return fetchGroup({
-                  id: selectedAnn.group,
-                  expand: params.expand,
-                }).then(directLinkedAnnGroup => {
-                  // If the directLinkedAnnotation's group fetch failed, return
-                  // the list of groups without it.
-                  if (!directLinkedAnnGroup) {
-                    return [
-                      combineGroups(myGroups, allFeaturedGroups, documentUri),
-                      token,
-                    ];
-                  }
-
-                  // If the directLinkedAnnotation's group fetch was successful,
-                  // combine it with the other groups.
-                  return [
-                    combineGroups(
-                      myGroups,
-                      allFeaturedGroups.concat(directLinkedAnnGroup),
-                      documentUri
-                    ),
-                    token,
-                  ];
-                });
+        return Promise.all(groupApiRequests);
+      })
+      .then(([myGroups, featuredGroups, token, selectedAnn, selectedGroup]) => {
+        // Step 3. Add the direct-linked group to the list of featured groups,
+        // and if there was a direct-linked annotation, fetch its group if we
+        // don't already have it.
+
+        // If there is a direct-linked group, add it to the featured groups list.
+        let allFeaturedGroups =
+          selectedGroup !== null &&
+          !featuredGroups.some(g => g.id === selectedGroup.id)
+            ? featuredGroups.concat([selectedGroup])
+            : featuredGroups;
+
+        // If there's a selected annotation it may require an extra api call
+        // to fetch its group.
+        if (selectedAnn) {
+          // Set the directLinkedAnnotationGroupId to be used later in
+          // the filterGroups method.
+          directLinkedAnnotationGroupId = selectedAnn.group;
+
+          const selectedAnnGroup = myGroups
+            .concat(allFeaturedGroups)
+            .some(g => g.id === selectedAnn.group);
+
+          // If the direct-linked annotation's group has not already been fetched,
+          // fetch it.
+          if (!selectedAnnGroup) {
```

Comment:
I find this Promise block challenging to parse. Maybe renaming the Boolean that indicates whether a group-fetch is needed and a slight restructuring something like the following?

```
if(weNeedToFetchTheGroup) { // obviously not a real name here

  const groupApi = fetchGroup({ /* ... */ });

  return groupApi.then(group => {
    if(group) {
      return allFeaturedGroups.concat(group);
    }
    return allFeaturedGroups;
  });
}
```

Specifically, I think the reassignment to `initialFeaturedGroups` feels confusing.